### PR TITLE
Nested object filtering — Part 13: DB-level test consolidation and same-K-different-parent fix

### PR DESCRIPTION
**Pull request by aliszka**
### Nested object filtering — Part 13                                                                                     
                                                                                                                            
  Thirteenth PR in the nested object filtering series, building on Part 12 (recursive plan/executor and same-element-at-LCA semantics). Ports the lower-level integration test matrix to the DB-level write+search pipeline, fixes a same-element correctness bug discovered during the port, and documents three remaining known-buggy behaviors that flip when later PRs land.                                                                                                                     
                                                                                                                          
  **Bug fix: same-K-different-parent**                                                                                      
   
  `runIdxLoopRecursive` uses `MaskLeafAnd` in `matchElementRecursive`, which strips leaf bits before ANDing per-condition bitmaps. When `_idx.{lcaPath}[K]` lumps multiple physical instances at the same K under different parents (e.g. `g[0].cars[0]` and `g[1].cars[0]`), `MaskLeaf` erases the only leaf-level distinction that would have rejected cross-instance matches.                                                                                                 
                                                                                                                          
  Fix adds a flat raw-AndAll path in `evalGroup`: when the subtree is a chain of `recGroupNode`s with unique `here` paths, no scalar-array terminals, and at most one deeper group with `here` conditions, fold the entire subtree into one raw `AndAll`. Same-element semantics holds via the analyzer's Phase-3 scalar inheritance (`elementPositions = descendants`), so leaves coincide within a single physical instance and diverge across instances. This avoids the expensive idx loop for filters where the encoding already enforces same-element semantics naturally.                                             

  The executor now receives the root prop's nested schema via `withProps` to identify scalar-array terminals (`text[]`, `int[]`, …), which use distinct leaves per element from `walkScalarArray` and break the inheritance bridge.
                                                                                                                            
  Six lower-level fixtures (`C3/C4/C5`, `F1/F3/F6`) were updated to assign leaves in DFS order per the walker — they had been forgiving under the `MaskLeaf` path but raw `AndAll` requires accurate leaves.
                                                                                                                            
  **DB-level test matrix port**                                                                                             
   
  Tests ported from the lower-level `TestResolveNestedCorrelatedAnd` / `TestCorrelatedAndFilterExamplesIndexed` / `TestPlanCasesIntegration` to the production write+search pipeline:                                                     
                                                                                                                            
  - **G1-G10 / C1-C10** correlated AND examples (each `Cn` mirrors `Gn` with `countries` prepended for two-depth coverage; gap-fills for unconstrained deep correlation, same-`arr[N]`-index on both sides, three-way multi-bucket splits)
  - **F1-F12, F14** four-condition indexed filter examples (forward and reverse orderings)                                  
  - **IsNull cluster** — standalone, IsNull-in-correlated-AND (28 sub-tests), IsNull+`arr[N]` at root/mid/deep, multi-level pinned IsNull, scalar-array positional IsNull, IsNull on multi-level `object[]`, `text[]` IsNull with `arr[N]`            
  - **`arr[N]` cluster** — positional access, level coverage (root/mid-1/mid-2), mixed constrained/unconstrained,           
  multi-level pin, three-clause overlap, partition vs union                                                                 
  - **Scalar-array positional** — basic, multi-level pin, partition with different parent indices, deeper-nested          
  (`cars.tires.tags[N]`), same-position contradiction                                                                       
  - **Tokenization follow-ups** — multi-token text + `arr[N]`, multi-token at deeper LCA (`cars.tires.*`), same-path      
  tokenized contradiction                                                                                                   
  - **Same-path multi-value** — tokenized and non-tokenized BUG ports (plain-object LCA × `object[]` LCA, `independent=1` vs `>1`)                                                                                                                    
  - **Comprehensive smoke test** — 28 sub-tests across both `DataTypeObject` and `DataTypeObjectArray` roots covering single-condition, AND, OR, parenthesized AND/OR mix, three-level deep nesting, and contradiction edge cases               
  - **Context cancellation** — four sub-tests exercising `canUseRawAndAll`, `runIdxLoopRecursive`, `rootAnchor`, and SPLIT executor entry paths                                                                                                      
                                                                                                                          
  **Negation regression coverage**                                                                                          
                                                                                                                          
  Three new test families lock in current behavior for negation on nested fields:                                           
  - `TestNestedFilteringOrOfCorrelatedAnds` — OR over two same-element AND groups
  - `TestNestedFilteringNotOfCorrelatedAnd` — NOT wrapping a correlated AND, with a regression sub-test pinning today's docID-level NOT semantics (flips to per-element in a later PR)                                                            
  - `TestNestedFilteringNotEqualNestedRegression` — locks in current universal `NotEqual` at top level and captures a discovered correctness bug: `NotEqual` inside a nested correlated AND is treated as `Equal` (denylist flag dropped); sub-test name flagged with `BUG` so the assertion is obviously locking buggy behavior                                   
                                                                                                                            
  **Test plan**                                                                                                           
  - [ ] `go test ./adapters/repos/db/inverted/...`
  - [ ] `go test -tags integrationTest ./adapters/repos/db/inverted/...`                                                    
  - [ ] `go test -tags integrationTest ./adapters/repos/db/...`

### CODE REVIEW

File: adapters/repos/db/inverted/prop_value_pairs_nested_integration_test.go [hunk 1]
```diff
@@ -6043,27 +6057,44 @@ func TestCorrelatedAndFilterExamples(t *testing.T) {
 		const (
 			docMatch   = uint64(1)
 			docNoMatch = uint64(2)
+
+			// docMatch data: countries[0].garages=[{city, cars:[{}, {make}]}].
+			// cars[0]/cars[1] each get a fresh leaf; garages[0].elementPositions
+			// is the union of cars's leaves; city inherits both.
```
[the remark this answers is not in the document]
Comment (aliszka) [reply adapters/repos/db/inverted/prop_value_pairs_nested_integration_test.go]: Fixed in `13656c9a29` — `cars's` → `cars'` across all 7 occurrences in the two test fixture files.

File: adapters/repos/db/inverted/prop_value_pairs_nested_integration_test.go [hunk 2]
```diff
@@ -6092,38 +6123,66 @@ func TestCorrelatedAndFilterExamples(t *testing.T) {
...
+			leafN2_G1 = uint16(2) // garages[1] (city only)
+			leafN2_C0 = uint16(3) // garages[2].cars[0]
+			leafN2_C1 = uint16(4) // garages[2].cars[1]
 		)
 		searcher, vb, mb := newFilterExamplesSearcher(t, "countries")
 
-		// docMatch: city and cars[1].make both in garages[2] of country[0]
-		writeNestedValue(t, vb, "garages.city", "berlin", []uint64{enc(1, leafG2Direct, docMatch)})
-		writeNestedValue(t, vb, "garages.cars.make", "honda", []uint64{enc(1, leafG1Cars1+2, docMatch)}) // garages[2].cars[1]
+		// docMatch: city inherits garages[2].cars's leaves; cars[1].make at cars[1].
```
[the remark this answers is not in the document]
Comment (aliszka) [reply adapters/repos/db/inverted/prop_value_pairs_nested_integration_test.go]: Fixed in `13656c9a29` — `cars's` → `cars'` across all 7 occurrences in the two test fixture files.

File: adapters/repos/db/inverted/prop_value_pairs_nested_integration_test.go [hunk 3]
```diff
@@ -6232,36 +6306,40 @@ func TestCorrelatedAndFilterExamplesIndexed(t *testing.T) {
...
+		writeNestedValue(t, vb, "garages.cars.make", "honda", []uint64{enc(1, leafG1C0, docMatch)})
+		writeNestedValue(t, vb, "garages.cars.model", "civic", []uint64{enc(1, leafG1C0, docMatch)})
 
-		// docNoMatch: same as docMatch but make in cars[0] and model in cars[1].
-		writeNestedValue(t, vb, "garages.city", "berlin", []uint64{enc(1, leafG0Direct, docNoMatch)})
-		writeNestedValue(t, vb, "garages.postcode", "12345", []uint64{enc(1, leafG1Direct, docNoMatch)})
-		writeNestedValue(t, vb, "garages.cars.make", "honda", []uint64{enc(1, leafG1Cars0, docNoMatch)})
-		writeNestedValue(t, vb, "garages.cars.model", "civic", []uint64{enc(1, leafG1Cars1, docNoMatch)})
+		// docNoMatch: same but garages[1].cars=[{make:honda}, {model:civic}].
+		// postcode inherits both cars's leaves.
```
[the remark this answers is not in the document]
Comment (aliszka) [reply adapters/repos/db/inverted/prop_value_pairs_nested_integration_test.go]: Fixed in `13656c9a29` — `cars's` → `cars'` across all 7 occurrences in the two test fixture files.

File: adapters/repos/db/inverted/prop_value_pairs_nested_integration_test.go [hunk 4]
```diff
@@ -6569,31 +6652,41 @@ func TestCorrelatedAndFilterExamplesIndexed(t *testing.T) {
 			docMatch   = uint64(1)
 			docNoMatch = uint64(2)
 
-			leafG0Direct = uint16(1)
-			leafG0Cars0  = uint16(2)
-			leafG0Cars1  = uint16(3)
+			// Per assign.go's walkObject: scalar fields inherit their
+			// parent element's positions, which equal the union of
+			// descendant leaves when descendants exist. So city/postcode
+			// at garages[0] inherit from cars's leaves; they do not get
```
[the remark this answers is not in the document]
Comment (aliszka) [reply adapters/repos/db/inverted/prop_value_pairs_nested_integration_test.go]: Fixed in `13656c9a29` — `cars's` → `cars'` across all 7 occurrences in the two test fixture files.

File: adapters/repos/db/inverted/prop_value_pairs_nested_integration_test.go [hunk 5]
```diff
@@ -6232,36 +6306,40 @@ func TestCorrelatedAndFilterExamplesIndexed(t *testing.T) {
 			docMatch   = uint64(1)
 			docNoMatch = uint64(2)
 
-			leafG0Direct = uint16(1)
-			leafG1Direct = uint16(2)
-			leafG1Cars0  = uint16(3)
-			leafG1Cars1  = uint16(4)
+			// Per assign.go's walkObject: postcode (a scalar on garages[1])
+			// inherits garages[1].elementPositions, which equals cars's
```
[the remark this answers is not in the document]
Comment (aliszka) [reply adapters/repos/db/inverted/prop_value_pairs_nested_integration_test.go]: Fixed in `13656c9a29` — `cars's` → `cars'` across all 7 occurrences in the two test fixture files.

File: adapters/repos/db/inverted/prop_value_pairs_nested_integration_test.go [hunk 6]
```diff
@@ -6569,31 +6652,41 @@ func TestCorrelatedAndFilterExamplesIndexed(t *testing.T) {
...
+
+		// docNoMatch: c[0].g[0]={city, postcode, cars:[{make}, {model}]}.
+		// Two cars → city/postcode inherit BOTH cars' leaves; make is at
+		// cars[0]'s leaf only; model is at cars[1]'s leaf only.
+		writeNestedValue(t, vb, "garages.city", "berlin", []uint64{enc(1, leafCar0, docNoMatch), enc(1, leafCar1, docNoMatch)})
+		writeNestedValue(t, vb, "garages.postcode", "12345", []uint64{enc(1, leafCar0, docNoMatch), enc(1, leafCar1, docNoMatch)})
+		writeNestedValue(t, vb, "garages.cars.make", "honda", []uint64{enc(1, leafCar0, docNoMatch)})
+		writeNestedValue(t, vb, "garages.cars.model", "civic", []uint64{enc(1, leafCar1, docNoMatch)})
+
+		// _idx[garages, 0]: union of cars's leaves (garages[0].elementPositions).
```
[the remark this answers is not in the document]
Comment (aliszka) [reply adapters/repos/db/inverted/prop_value_pairs_nested_integration_test.go]: Fixed in `13656c9a29` — `cars's` → `cars'` across all 7 occurrences in the two test fixture files.

File: adapters/repos/db/inverted/searcher_nested_executor_recursive.go [hunk 7]
```diff
@@ -228,20 +245,177 @@ func (e *recExecutor) evalNode(ctx context.Context, node recPlanNode, parentScop
...
+	}
+	if !walk(g, true) {
+		return nil, false
+	}
+	return out, true
+}
+
+// pathTerminalIsScalarArray reports whether path's terminal property in the
+// nested schema is a scalar-array type (text[], int[], number[], boolean[],
+// date[]).
```
[the remark this answers is not in the document]
Comment (aliszka) [reply adapters/repos/db/inverted/searcher_nested_executor_recursive.go]: Fixed in `13656c9a29` — added `uuid[]` to the three scalar-array doc-comment enumerations. Runtime detection already uses `schema.IsScalarArrayType` which includes `DataTypeUUIDArray`; this was just doc alignment.

File: adapters/repos/db/inverted/prop_value_pairs_nested_integration_test.go [hunk 8]
```diff
@@ -6043,27 +6057,44 @@ func TestCorrelatedAndFilterExamples(t *testing.T) {
...
+		// _idx[garages, 1]: only docNoMatch has garages[1].
 		require.NoError(t, mb.RoaringSetAddList(invnested.IdxKey("garages", 1),
-			[]uint64{enc(1, leafG1Cars1, docNoMatch)}))
-		// _idx.cars[1]: positions of cars[1] in all garages
+			[]uint64{enc(1, leafNoMatchC0, docNoMatch), enc(1, leafNoMatchC1, docNoMatch)}))
+		// _idx[cars, 0]: cars[0] across docs.
+		require.NoError(t, mb.RoaringSetAddList(invnested.IdxKey("cars", 0),
+			[]uint64{enc(1, leafMatchC0, docMatch), enc(1, leafNoMatchC0, docNoMatch)}))
+		// _idx[cars, 1]: cars[1] across docs.
 		require.NoError(t, mb.RoaringSetAddList(invnested.IdxKey("cars", 1),
```
[the remark this answers is not in the document]
Comment (aliszka) [reply adapters/repos/db/inverted/prop_value_pairs_nested_integration_test.go]: Good catch — fixed in `13656c9a29`. Three flavors: C2/C3/C4 (filter pins `cars[N]`) had both key and `ArrayIndex.RelPath` corrected to `garages.cars` (the wrong `RelPath: "cars"` was silently dropping the pin so tests passed via raw-leaf-alignment); C5/F1-F6 (no `cars[N]` pin) had orphaned rows deleted since dispatch goes through `canUseRawAndAll`/`evalFlatRawAndAll`; F7/F8 (scalar-array terminal or deeper sub-arrays force `runIdxLoopRecursive`) had rows renamed to `IdxKey("garages.cars", N)` with an explanatory comment.

File: adapters/repos/db/inverted/prop_value_pairs_nested_integration_test.go [hunk 9]
```diff
@@ -6151,27 +6210,42 @@ func TestCorrelatedAndFilterExamples(t *testing.T) {
...
 		require.NoError(t, mb.RoaringSetAddList(invnested.IdxKey("garages", 0),
 			[]uint64{
-				enc(1, leafG0Direct, docMatch), enc(1, leafG0Cars1, docMatch),
-				enc(1, leafG0Direct, docNoMatch), enc(1, leafG0Cars0, docNoMatch), enc(1, leafG0Cars1, docNoMatch),
+				enc(1, leafM_C0, docMatch), enc(1, leafM_C1, docMatch),
+				enc(1, leafN_C0, docNoMatch), enc(1, leafN_C1, docNoMatch),
 			}))
+		require.NoError(t, mb.RoaringSetAddList(invnested.IdxKey("cars", 0),
+			[]uint64{enc(1, leafM_C0, docMatch), enc(1, leafN_C0, docNoMatch)}))
+		require.NoError(t, mb.RoaringSetAddList(invnested.IdxKey("cars", 1),
```
[the remark this answers is not in the document]
Comment (aliszka) [reply adapters/repos/db/inverted/prop_value_pairs_nested_integration_test.go]: Good catch — fixed in `13656c9a29`. Three flavors: C2/C3/C4 (filter pins `cars[N]`) had both key and `ArrayIndex.RelPath` corrected to `garages.cars` (the wrong `RelPath: "cars"` was silently dropping the pin so tests passed via raw-leaf-alignment); C5/F1-F6 (no `cars[N]` pin) had orphaned rows deleted since dispatch goes through `canUseRawAndAll`/`evalFlatRawAndAll`; F7/F8 (scalar-array terminal or deeper sub-arrays force `runIdxLoopRecursive`) had rows renamed to `IdxKey("garages.cars", N)` with an explanatory comment.

File: adapters/repos/db/inverted/prop_value_pairs_nested_integration_test.go [hunk 10]
```diff
@@ -6359,45 +6437,50 @@ func TestCorrelatedAndFilterExamplesIndexed(t *testing.T) {
...
+				enc(1, leafC0G0, docMatch), enc(1, leafC0G0, docNoMatch),
+				enc(2, leafC1G0C0, docMatch),
+				enc(2, leafC1G0C0, docNoMatch), enc(2, leafC1G0C1, docNoMatch),
 			}))
 		require.NoError(t, mb.RoaringSetAddList(invnested.IdxKey("cars", 0),
-			[]uint64{enc(2, leafC1G0Cars0, docMatch), enc(2, leafC1G0Cars0, docNoMatch)}))
+			[]uint64{enc(2, leafC1G0C0, docMatch), enc(2, leafC1G0C0, docNoMatch)}))
 		require.NoError(t, mb.RoaringSetAddList(invnested.IdxKey("cars", 1),
-			[]uint64{enc(2, leafC1G0Cars1, docNoMatch)}))
+			[]uint64{enc(2, leafC1G0C1, docNoMatch)}))
```
[the remark this answers is not in the document]
Comment (aliszka) [reply adapters/repos/db/inverted/prop_value_pairs_nested_integration_test.go]: Good catch — fixed in `13656c9a29`. Three flavors: C2/C3/C4 (filter pins `cars[N]`) had both key and `ArrayIndex.RelPath` corrected to `garages.cars` (the wrong `RelPath: "cars"` was silently dropping the pin so tests passed via raw-leaf-alignment); C5/F1-F6 (no `cars[N]` pin) had orphaned rows deleted since dispatch goes through `canUseRawAndAll`/`evalFlatRawAndAll`; F7/F8 (scalar-array terminal or deeper sub-arrays force `runIdxLoopRecursive`) had rows renamed to `IdxKey("garages.cars", N)` with an explanatory comment.

File: adapters/repos/db/inverted/prop_value_pairs_nested_integration_test.go [hunk 11]
```diff
@@ -6569,31 +6652,41 @@ func TestCorrelatedAndFilterExamplesIndexed(t *testing.T) {
...
-				enc(1, leafG0Direct, docNoMatch), enc(1, leafG0Cars0, docNoMatch), enc(1, leafG0Cars1, docNoMatch),
+				enc(1, leafCar0, docMatch),
+				enc(1, leafCar0, docNoMatch), enc(1, leafCar1, docNoMatch),
 			}))
 		require.NoError(t, mb.RoaringSetAddList(invnested.IdxKey("cars", 0),
-			[]uint64{enc(1, leafG0Cars0, docMatch), enc(1, leafG0Cars0, docNoMatch)}))
+			[]uint64{enc(1, leafCar0, docMatch), enc(1, leafCar0, docNoMatch)}))
 		require.NoError(t, mb.RoaringSetAddList(invnested.IdxKey("cars", 1),
-			[]uint64{enc(1, leafG0Cars1, docNoMatch)}))
+			[]uint64{enc(1, leafCar1, docNoMatch)}))
```
[the remark this answers is not in the document]
Comment (aliszka) [reply adapters/repos/db/inverted/prop_value_pairs_nested_integration_test.go]: Good catch — fixed in `13656c9a29`. Three flavors: C2/C3/C4 (filter pins `cars[N]`) had both key and `ArrayIndex.RelPath` corrected to `garages.cars` (the wrong `RelPath: "cars"` was silently dropping the pin so tests passed via raw-leaf-alignment); C5/F1-F6 (no `cars[N]` pin) had orphaned rows deleted since dispatch goes through `canUseRawAndAll`/`evalFlatRawAndAll`; F7/F8 (scalar-array terminal or deeper sub-arrays force `runIdxLoopRecursive`) had rows renamed to `IdxKey("garages.cars", N)` with an explanatory comment.

File: adapters/repos/db/inverted/prop_value_pairs_nested_integration_test.go [hunk 12]
```diff
@@ -6092,38 +6123,66 @@ func TestCorrelatedAndFilterExamples(t *testing.T) {
...
-		// _idx.cars[1]: cars[1] positions across all garages
+		// _idx.cars[1]: cars[1] positions across all garages.
 		require.NoError(t, mb.RoaringSetAddList(invnested.IdxKey("cars", 1),
 			[]uint64{
-				enc(1, leafG1Cars1+2, docMatch), enc(1, leafG1Cars1, docNoMatch1),
-				enc(1, leafG1Cars1+2, docNoMatch2),
+				enc(1, leafM_C1, docMatch),
+				enc(1, leafN1_C1, docNoMatch1),
+				enc(1, leafN2_C1, docNoMatch2),
 			}))
```
[the remark this answers is not in the document]
Comment (aliszka) [reply adapters/repos/db/inverted/prop_value_pairs_nested_integration_test.go]: Good catch — fixed in `13656c9a29`. Three flavors: C2/C3/C4 (filter pins `cars[N]`) had both key and `ArrayIndex.RelPath` corrected to `garages.cars` (the wrong `RelPath: "cars"` was silently dropping the pin so tests passed via raw-leaf-alignment); C5/F1-F6 (no `cars[N]` pin) had orphaned rows deleted since dispatch goes through `canUseRawAndAll`/`evalFlatRawAndAll`; F7/F8 (scalar-array terminal or deeper sub-arrays force `runIdxLoopRecursive`) had rows renamed to `IdxKey("garages.cars", N)` with an explanatory comment.

File: adapters/repos/db/inverted/prop_value_pairs_nested_integration_test.go [hunk 13]
```diff
@@ -6232,36 +6306,40 @@ func TestCorrelatedAndFilterExamplesIndexed(t *testing.T) {
...
-				enc(1, leafG1Direct, docNoMatch), enc(1, leafG1Cars0, docNoMatch), enc(1, leafG1Cars1, docNoMatch),
+				enc(1, leafG1C0, docMatch),
+				enc(1, leafG1C0, docNoMatch), enc(1, leafG1C1, docNoMatch),
 			}))
 		require.NoError(t, mb.RoaringSetAddList(invnested.IdxKey("cars", 0),
-			[]uint64{enc(1, leafG1Cars0, docMatch), enc(1, leafG1Cars0, docNoMatch)}))
+			[]uint64{enc(1, leafG1C0, docMatch), enc(1, leafG1C0, docNoMatch)}))
 		require.NoError(t, mb.RoaringSetAddList(invnested.IdxKey("cars", 1),
-			[]uint64{enc(1, leafG1Cars1, docNoMatch)}))
+			[]uint64{enc(1, leafG1C1, docNoMatch)}))
```
[the remark this answers is not in the document]
Comment (aliszka) [reply adapters/repos/db/inverted/prop_value_pairs_nested_integration_test.go]: Good catch — fixed in `13656c9a29`. Three flavors: C2/C3/C4 (filter pins `cars[N]`) had both key and `ArrayIndex.RelPath` corrected to `garages.cars` (the wrong `RelPath: "cars"` was silently dropping the pin so tests passed via raw-leaf-alignment); C5/F1-F6 (no `cars[N]` pin) had orphaned rows deleted since dispatch goes through `canUseRawAndAll`/`evalFlatRawAndAll`; F7/F8 (scalar-array terminal or deeper sub-arrays force `runIdxLoopRecursive`) had rows renamed to `IdxKey("garages.cars", N)` with an explanatory comment.

File: adapters/repos/db/inverted/searcher_nested_plan_recursive.go [hunk 14]
```diff
@@ -170,22 +170,78 @@ func (b *recPlanBuilder) buildGroup(items []*propValuePair, scope string) recPla
...
+// by root_idx in the position encoding, no outer loop needed.
 func needsWrappingGroup(scope string, sub recPlanNode) bool {
 	if scope == "" {
 		return false
 	}
-	split, ok := sub.(*recSplitNode)
-	if !ok {
+	switch n := sub.(type) {
+	case *recSplitNode:
+		return len(n.branches) > 1
```
[the remark this answers is not in the document]
Comment (aliszka) [reply adapters/repos/db/inverted/searcher_nested_plan_recursive.go]: Fixed in `aa9368cb62` — `needsWrappingGroup`'s `recSplitNode` case now recurses into the single branch's plan, so a 1-branch SPLIT containing a descendant that needs outer scope keeps the wrap. New regression test `two scalar-array siblings under 1-branch SPLIT — wrapping group preserves per-garage scope` fails before, passes after.

File: adapters/repos/db/inverted/searcher_nested_plan_recursive.go [hunk 15]
```diff
@@ -170,22 +170,78 @@ func (b *recPlanBuilder) buildGroup(items []*propValuePair, scope string) recPla
...
+//   - It has ≥2 subs (the flat raw-AndAll path bails on multi-sub since
+//     sibling sub-arrays produce conditions at distinct leaves with no
+//     inheritance bridge).
+//   - It has duplicate here paths (same-path values land at distinct leaves
+//     per walkScalarArray, so canUseRawAndAll is false and the flat path
+//     bails). Same-element semantics for these requires per-element _idx
+//     iteration scoped by the outer parent.
+func groupSubtreeNeedsOuterScope(g *recGroupNode) bool {
+	if g.lcaPath != "" && (len(g.subs) >= 2 || hasDuplicateHerePaths(g)) {
+		return true
```
[the remark this answers is not in the document]
Comment (aliszka) [reply adapters/repos/db/inverted/searcher_nested_plan_recursive.go:228]: Fixed in `aa9368cb62` — new `groupNeedsIdxLoop` helper mirrors `collectFlatSubtree`'s scalar-array bails (≥2 scalar-array here, or ≥1 scalar-array + ≥1 sub); the >1-deeper-here-chain gap requires a contrived ≥3-level structure and isn't fixed here. Regression test `two scalar-array siblings at intermediate LCA — same-car-across-garages disambiguation` fails before, passes after.

File: adapters/repos/db/inverted/searcher_nested_executor_recursive.go [hunk 16]
```diff
@@ -228,20 +245,177 @@ func (e *recExecutor) evalNode(ctx context.Context, node recPlanNode, parentScop
...
+//     sub plans must all match within the same element.
 //
 // The here=0, subs=1 case is collapsed away by the planner so it does not
 // reach evalGroup (see buildGroup).
 func (e *recExecutor) evalGroup(ctx context.Context, g *recGroupNode, parentScope *sroar.Bitmap) (*sroar.Bitmap, func(), error) {
 	if canUseRawAndAll(g) || g.lcaPath == "" {
 		return e.evalGroupRoot(ctx, g, parentScope)
 	}
+	if flat, ok := e.collectFlatSubtree(g); ok {
+		return e.evalFlatRawAndAll(ctx, flat, parentScope)
```
[the remark this answers is not in the document]
Comment (aliszka) [reply adapters/repos/db/inverted/searcher_nested_executor_recursive.go]: Fixed in `aa9368cb62` — `canUseRawAndAll` is now a method on `recExecutor` and bails to `runIdxLoopRecursive` when ≥2  scalar-array siblings are present (a single scalar-array + regular scalars still works via Phase-3 inheritance).         
  Regression test `two scalar-array siblings — single value each — same-car semantics` fails before, passes after.

File: adapters/repos/db/inverted/searcher_nested_plan_recursive.go [hunk 17]
```diff
@@ -170,22 +170,143 @@ func (b *recPlanBuilder) buildGroup(items []*propValuePair, scope string) recPla
...
+		if i == len(segs)-1 {
+			return schema.IsScalarArrayType(dt)
+		}
+		if !schema.IsNested(dt) {
+			return false
+		}
+		props = np.NestedProperties
+	}
+	return false
+}
```
[the remark this answers is not in the document]
Comment (aliszka) [reply adapters/repos/db/inverted/searcher_nested_plan_recursive.go:291]: fixed

File: adapters/repos/db/inverted/searcher_nested_plan_recursive.go [hunk 18]
```diff
@@ -170,22 +170,143 @@ func (b *recPlanBuilder) buildGroup(items []*propValuePair, scope string) recPla
...
+
+// hasDuplicateHerePaths reports whether two or more here entries in g share
+// the same childRelPath. Each text[]/scalar-array value lives at a distinct
+// leaf assigned by walkScalarArray, so duplicate-path filters cannot match
+// at the same leaf via raw AndAll — they need per-element evaluation.
```
[the remark this answers is not in the document]
Comment (aliszka) [reply adapters/repos/db/inverted/searcher_nested_plan_recursive.go:296]: fixed

File: adapters/repos/db/inverted/searcher_nested_executor_recursive.go [hunk 19]
```diff
@@ -228,20 +245,177 @@ func (e *recExecutor) evalNode(ctx context.Context, node recPlanNode, parentScop
...
+//   - empty props (without schema we cannot identify scalar-array paths;
+//     stay on runIdxLoopRecursive)
+//   - recSplitNode anywhere (arr[N] requires per-element evaluation)
+//   - duplicate here paths (each value lives at a distinct leaf — raw AndAll
+//     over duplicates over-rejects, even within the same physical element)
+//   - any group with ≥2 subs (sibling sub-arrays produce conditions at
+//     distinct leaves with no inheritance bridge — raw AndAll fails)
+//   - any here condition on a scalar-array (text[], int[], uuid[], …) path.
+//     These values are written by walkScalarArray with one leaf per element
+//     (assign.go), so they don't share leaves via Phase-3 inheritance with
```
[the remark this answers is not in the document]
Comment (aliszka) [reply adapters/repos/db/inverted/searcher_nested_executor_recursive.go]: fixed

File: adapters/repos/db/inverted/searcher_nested_executor_recursive.go [hunk 20]
```diff
@@ -329,23 +503,40 @@ func (e *recExecutor) evalGroupRoot(ctx context.Context, g *recGroupNode, parent
...
+// MaskLeafAnd preserves correctness by comparing rootDoc after stripping
+// leaves.
+func (e *recExecutor) canUseRawAndAll(g *recGroupNode) bool {
 	if len(g.subs) > 0 || len(g.here) == 0 {
 		return false
 	}
 	if len(g.here) == 1 {
 		return true
 	}
 	seen := make(map[string]struct{}, len(g.here))
```
[the remark this answers is not in the document]
Comment (aliszka) [reply adapters/repos/db/inverted/searcher_nested_executor_recursive.go:508]: fixed

**Review by amourao — DISMISSED** [review]

**Review by antas-marcin — APPROVED** [review]

**Review by antas-marcin — APPROVED** [review]
